cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java

```diff
@@ -181,7 +181,9 @@ public void run() {
    * Children call this method to signal the program is completed.
    */
   protected void complete() {
-    if (!state.compareAndSet(State.ALIVE, State.COMPLETED)) {
+    if (!state.compareAndSet(State.STARTING, State.COMPLETED)
```

Comment:
I think the `AbstractProgramController` need some revamp. The current implementation is not good and it would be much simpler if we just use lock for state changes. I'll leave the change as is for now since it will involve rewriting the class.